src/Tokenizer/Tokens.php

```diff
@@ -337,11 +340,11 @@ public function clearEmptyTokens(): void
 
         for ($count = $index; $index < $limit; ++$index) {
             if (!$this->isEmptyAt($index)) {
-                $this[$count++] = $this[$index]; // @phpstan-ignore-line as we know that index exists
+                // use directly for speed, skip the register of token kinds found etc.
+                parent::offsetSet($count++, $this[$index]);
             }
         }
 
-        // we are moving the tokens, we need to clear the indices Cache
         $this->blockStartCache = [];
```

Comment:
Your initial change to remove the cache clearing was valid and working because the cache gets cleared on "offset".
However, if we move a lot of tokens than we would save a lot of cycles by doing the `offset` directly through the parent, skipping the multiple cache clear class (we really need only one at the end of this method) and we skip all the token types found updates.

Reply:
nice optimisation!